File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/binary_tree/P222_CountCompleteTreeNodes.java
package com.dsa.leetcode.labuladong.binary_tree;

public class P222_CountCompleteTreeNodes {

    public int countNodes(TreeNode root) {
        if (root == null) {
            return 0;
        }
        int left = countHeight(root.left);
        int right = countHeight(root.right);
        if (left == right) {
            return countNodes(root.right) + (1 << left);
        } else {
            return countNodes(root.left) + (1 << right);
        }
    }

    private int countHeight(TreeNode root) {
        int h = 0;
        while (root != null) {
            h++;
            root = root.left;
        }
        return h;
    }
}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/binary_tree/P222_CountCompleteTreeNodes.py
# author: tobebetter9527
# since: 2024/4/24 8:58
from typing import Optional

from com.dsa.leetcode.labuladong.binary_tree import TreeNode


class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0
        left = self.countHeight(root.left)
        right = self.countHeight(root.right)
        if left == right:
            return self.countNodes(root.right) + (1 << left)
        else:
            return self.countNodes(root.left) + (1 << right)

    def countHeight(self, root):
        h = 0
        while root:
            h += 1
            root = root.left
        return h
